Unit testing in .NET blog post (#3343)
pulumi/docs: content/blog/unit-testing-cloud-deployments-with-dotnet/index.md

```

{{% /choosable %}}

I defined a helper method `TestAsync` that points Pulumi's `Deployment.TestAsync` to our stack and mock classes.

As we progress through the article, I will add tests to this class one-by-one.

### Retrieve output values

Finally, I need a small extension method to extract values from outputs. Every Pulumi resource returns values wrapped inside an `Output<T>` container. While the real engine runs, those values may be known or unknown, but my mock-based tests always return known values. It's safe to get those values and convert them to a `Task<T>`:

{{< chooser language "csharp,fsharp" />}}

{{% choosable language csharp %}}

```csharp
public static class TestingExtensions
{
    public static Task<T> GetValueAsync<T>(this Output<T> output)
    {
        var tcs = new TaskCompletionSource<T>();
        output.Apply(v => { tcs.SetResult(v); return v; });
        return tcs.Task;
    }
}
```

{{% /choosable %}}
{{% choosable language fsharp %}}

```fsharp
let getValue(output: Output<'a>): 'a =
    let tcs = new TaskCompletionSource<'a>()
    output.Apply(fun v -> tcs.SetResult(v); v) |> ignore
    tcs.Task.Result
```

{{% /choosable %}}

> To learn more about outputs, read [this article](/docs/intro/concepts/programming-model/#stack-outputs).

Reviewer: ```suggestion
To learn more about outputs, read [this article]({{< relref "/docs/intro/concepts/programming-model#stack-outputs" >}}).
```

{{< chooser language "csharp,fsharp" />}}

{{% choosable language csharp %}}

```csharp
[Test]
public async Task StorageAccountBelongsToResourceGroup()
{
    var resources = await TestAsync();
    var storageAccount = resources.OfType<Storage.Account>().SingleOrDefault();
    storageAccount.Should().NotBeNull("Storage account not found");

    var resourceGroupName = await storageAccount.ResourceGroupName.GetValueAsync();
    resourceGroupName.Should().Be("www-prod-rg");
}
```

{{% /choosable %}}
{{% choosable language fsharp %}}

```fsharp
[<Test>]
member this.StorageAccountBelongsToResourceGroup() =
    let resources = runTest()
    let storageAccount = resources.OfType<Account>() |> Seq.tryHead |> Option.toObj
    storageAccount.Should().NotBeNull("Storage account not found") |> ignore

    let resourceGroupName = getValue storageAccount.ResourceGroupName
    resourceGroupName.Should().Be("www-prod-rg", null) |> ignore
```

{{% /choosable %}}

> Of course, the test fails. I go ahead and try to fix it with what I think is the minimal change to make the test pass: I add a new resource to the stack and point it to the resource group.

Reviewer: ```suggestion
Of course, the test fails. I try to fix it with what I think is the minimal change to make the test pass by adding a new resource to the stack and pointing it to the resource group.
```

{{< chooser language "csharp,fsharp" />}}

{{% choosable language csharp %}}

```csharp
var storageAccount = new Storage.Account("wwwprodsa", new Storage.AccountArgs
{
    ResourceGroupName = resourceGroup.Name,
    AccountTier = "Standard",
    AccountReplicationType = "LRS",
    StaticWebsite = new Storage.Inputs.AccountStaticWebsiteArgs
    {
        IndexDocument = "index.html"
    }
});
```

{{% /choosable %}}
{{% choosable language fsharp %}}

```fsharp
let storageAccount =
    new Account(
        "wwwprodsa",
        new AccountArgs(
            ResourceGroupName = io resourceGroup.Name,
            AccountTier = input "Standard",
            AccountReplicationType = input "LRS",
            StaticWebsite = input (new AccountStaticWebsiteArgs(IndexDocument = input "index.html"))))
```

{{% /choosable %}}

I defined the `StaticWebsite` property as well: I'll leave testing these values as an exercise for the reader.

Two resource group tests are back to green again, but the storage account test fails with a new error message.

```
X StorageAccountBelongsToResourceGroup [84ms]
  Error Message:
   Expected resourceGroupName to be "www-prod-rg", but found <null>.
```

> What's going on, why is it `null`?

Reviewer: ```suggestion
What's going on, and why is it `null`?
```

I defined the account name like this: `ResourceGroupName = resourceGroup.Name`. However, if we look closely, the `resourceGroup` resource doesn't have an input property `Name` defined. `www-prod-rg` is a logical name for Pulumi deployment, not the physical name of the resource.

> Under normal circumstances, the Pulumi engine would use the logical name to produce the physical name of the resource group automatically (see [resource names](https://www.pulumi.com/docs/intro/concepts/programming-model/#names) for details). However, my mocks don't do that.

Reviewer: ```suggestion
Under normal circumstances, the Pulumi engine would use the logical name to produce the physical name of the resource group automatically (see [resource names]({{< relref "/docs/intro/concepts/programming-model#names" >}}) for details). However, my mocks don't do that.
```

{{< chooser language "csharp,fsharp" />}}

{{% choosable language csharp %}}

```csharp
// Set the name to resource name if it's not set explicitly in inputs.
if (!inputs.ContainsKey("name"))
    outputs.Add("name", name);
```

{{% /choosable %}}
{{% choosable language fsharp %}}

```fsharp
let inputKVs = inputs |> Seq.map(fun kv -> kv.Key, kv.Value) |> Seq.toList
let nameKVs = if inputs.ContainsKey("name") then [] else [("name", name :> obj)]

let outputs = [inputKVs; nameKVs] |> Seq.concat |> Seq.map KeyValuePair
let dict = outputs.ToImmutableDictionary() :> obj
```

{{% /choosable %}}

Note that mocks operate on weakly-typed dictionaries, so I need to get the property name right. Pulumi SDKs are open source, so I looked [here](https://github.com/pulumi/pulumi-azure/blob/1fdcab88065175ced768d900e7dcedf3b1d1b0a7/sdk/dotnet/Core/ResourceGroup.cs#L90) to double-check the exact value.

After this change in `Mocks`, my tests go green again.

```
Test Run Successful.
Total tests: 3
     Passed: 3
```

## Website Files

The next step is to upload some files to the static website. Well, instead, to write an automated test that validates the upload with mocks.

> I create a `wwwroot` folder with two HTML files in it and configure to copy that folder to the output on the build. Here is my change in the project file.

Reviewer: ```suggestion
I create a `wwwroot` folder with two HTML files in it and copy the folder to the build's output. Here is my change in the project file.
```

foreach (var file in files)
{
    var blob = new Storage.Blob(file, new Storage.BlobArgs
    {
        ContentType = "application/html",
        Source = new FileAsset(file),
        StorageAccountName = storageAccount.Name,
        StorageContainerName = "$web",
        Type = "Block"
    });
}
```

{{% /choosable %}}
{{% choosable language fsharp %}}

```fsharp
let files =
    let fileNames = Directory.GetFiles("wwwroot")
    fileNames
    |> Seq.map(fun file ->
        new Blob(
            file,
            new BlobArgs(
                ContentType = input "application/html",
                Source = input (new FileAsset(file) :> AssetOrArchive),
                StorageAccountName = io storageAccount.Name,
                StorageContainerName = input "$web",
                Type = input "Block")))
    |> List.ofSeq
```

{{% /choosable %}}

However, this change breaks the stack and all tests in the suite:

```
Error Message:
   Pulumi.RunException : Running program failed with an unhandled exception:
System.InvalidOperationException: Unsupported value when converting to protobuf: Pulumi.FileAsset
   at Pulumi.Serialization.Serializer.CreateValue(Object value)
...
```

> Once again, our mocks fail to represent the behavior of the engine accurately. The engine knows about the `FileAsset` class that points to a file on the disk and how to convert it to an uploaded blob. Also, the engine doesn't copy this property to outputs. I need to adjust the mocks again.

Reviewer: ```suggestion
Once again, our mocks fail to represent the behavior of the engine accurately. The Pulumi engine knows about the FileAsset class pointing to a file on the disk and how to convert it to an uploaded blob. But, the engine doesn't copy this property to outputs. I need to adjust the mocks again.
```
